design-editor/src/pane/style-manager.js

```diff
@@ -58,28 +58,97 @@ class StyleManager {
+	 * @param {string | object} newStyleDefinition
+	 * @private
+	 */
+	_onChangeElementStyle(newStyleDefinition) {
+		// eslint-disable-next-line no-console
+		console.log('style-manager:_onChangeElementStyle');
+		const id = elementSelector.getSelectedElementId(),
+			designEditor = appManager.getActiveDesignEditor(),
+			$element = designEditor && $(designEditor._getElementById(id));
+		let template;
 
-        var id = elementSelector.getSelectedElementId(),
-            designEditor = appManager.getActiveDesignEditor(),
-            $element = designEditor && $(designEditor._getElementById(id));
+		this._designEditor = designEditor;
+		if (designEditor.getUIInfo($element).package.name === 'listview') {
+			this._getModel($element.children(), 'listview');
+		} else {
+			this._getModel($element, designEditor.getUIInfo($element).package.name);
+		}
+		if (typeof newStyleDefinition === 'string') {
+			template = newStyleDefinition;
+			this._replaceStyle($element, template);
+		} else {
+			this._applyStyle($element, newStyleDefinition);
+		}
+	}
 
-        this._designEditor = designEditor;
+	/**
+	 * Applies defined style for element
+	 * @param {jQuery} $element
+	 * @param {Object} newStyleDefinition
+	 */
+	_applyStyle($element, newStyleDefinition) {
+		const id = $element.attr('data-id'),
+			model = this._designEditor.getModel(),
+			element = model.getElement(id),
+			component = element && element.component,
+			currentClassList = element.classList,
+			newClassList = newStyleDefinition.classList || [],
+			newAttributes = newStyleDefinition.attributes || {},
+			newStyles = newStyleDefinition.style || {},
+			classesSet = new Set(); // classes used across widget available templates
 
-        if (designEditor.getUIInfo($element).package.name === 'listview') {
-            this._getModel($element.children(), 'listview');
-        } else {
-            this._getModel($element, designEditor.getUIInfo($element).package.name);
-        }
+		// Looking for classes that should be removed.
+		// To avoid removing unnecessary classes search for
+		// all classes used across widget available templates.
+		// Leave classes that aren't mentioned in any template
+		// or classes are defined in current preset style.
+		if (component && component.options && component.options.styles) {
+			component.options.styles.forEach(newStyleDefinition => {
+				// For compatibility with previous string style templates
+				if (typeof newStyleDefinition !== 'string' &&
+					newStyleDefinition.template &&
+					newStyleDefinition.template.classList) {
+					newStyleDefinition.template.classList.forEach(c => classesSet.add(c));
+				}
+			});
+		}
 
-        this._replaceStyle($element, template);
-    }
+		// Remove unnecessary classes
+		classesSet.forEach((className) => {
+			const index = currentClassList.indexOf(className);
+			if (index !== -1) {
+				currentClassList.splice(index, 1);
+			}
+		});
+
+		currentClassList.push(...newClassList);
```

Comment:
I don't understand why you use `push` method with `...` instead of `concat`. Does the variable "currentClassList" have to be the same instance?

Reply:
I have changed the implementation